### PR TITLE
[DC-AE] support tiling for DC-AE

### PR DESCRIPTION
**Pull request by chenjy2003**
# What does this PR do?

This PR will add tiling support for DC-AE (Deep Compression Autoencoder for Efficient High-Resolution Diffusion Models) into the diffusers lib, in order to reduce memory consumption for encoding and decoding super-high-resolution images like 4096x4096. 

Could you please have a look? Thanks! @sayakpaul @yiyixuxu 
Cc: @lawrence-cj

### CODE REVIEW

**Comment by elismasilva** [comment]
After this PR merged, It will be necessary to implement StableDiffusionMixIn inheritance in SANA pipelines in order to enable tiling and slicing.

File: src/diffusers/models/autoencoders/autoencoder_dc.py [hunk 1]
```diff
@@ -479,12 +479,15 @@ def __init__(
         self.use_tiling = False
 
         # The minimal tile height and width for spatial tiling to be used
-        self.tile_sample_min_height = 512
-        self.tile_sample_min_width = 512
+        self.tile_sample_min_height = 1024
```
Comment (a-r-r-o-w) [inline src/diffusers/models/autoencoders/autoencoder_dc.py:482]: This change should be safe to make as DCAE is already quite low-memory usage with default settings

File: src/diffusers/models/autoencoders/autoencoder_dc.py [hunk 2]
```diff
@@ -606,11 +611,101 @@ def decode(self, z: torch.Tensor, return_dict: bool = True) -> Union[DecoderOutp
...
+
+        # Split x into overlapping tiles and encode them separately.
+        # The tiles have an overlap to avoid seams between tiles.
+        rows = []
+        for i in range(0, x.shape[2], self.tile_sample_stride_height):
+            row = []
+            for j in range(0, x.shape[3], self.tile_sample_stride_width):
+                tile = x[:, :, i : i + self.tile_sample_min_height, j : j + self.tile_sample_min_width]
+                if tile.shape[2] % self.spatial_compression_ratio != 0 or tile.shape[3] % self.spatial_compression_ratio != 0:
+                    tile = F.pad(tile, (0, (self.spatial_compression_ratio - tile.shape[3]) % self.spatial_compression_ratio, 0, (self.spatial_compression_ratio - tile.shape[2]) % self.spatial_compression_ratio))
```
Comment (a-r-r-o-w) [inline src/diffusers/models/autoencoders/autoencoder_dc.py:646]: (nit): Would create variables for padding length here because the line becomes too long

**Review by a-r-r-o-w — COMMENTED** [review]
Thanks @lawrence-cj! The changes look good.

As mentioned by the @elismasilva, we need to add some methods to the pipeline. Let's not derive from `StableDiffusionMixin` however, since it contains `enable_freeu` and `disable_freeu`.

Would just copy this into the pipeline:

```python
    def enable_vae_slicing(self):
        r"""
        Enable sliced VAE decoding. When this option is enabled, the VAE will split the input tensor in slices to
        compute decoding in several steps. This is useful to save some memory and allow larger batch sizes.
        """
        self.vae.enable_slicing()

    def disable_vae_slicing(self):
        r"""
        Disable sliced VAE decoding. If `enable_vae_slicing` was previously enabled, this method will go back to
        computing decoding in one step.
        """
        self.vae.disable_slicing()

    def enable_vae_tiling(self):
        r"""
        Enable tiled VAE decoding. When this option is enabled, the VAE will split the input tensor into tiles to
        compute decoding and encoding in several steps. This is useful for saving a large amount of memory and to allow
        processing larger images.
        """
        self.vae.enable_tiling()

    def disable_vae_tiling(self):
        r"""
        Disable tiled VAE decoding. If `enable_vae_tiling` was previously enabled, this method will go back to
        computing decoding in one step.
        """
        self.vae.disable_tiling()
```

**Comment by FurkanGozukara** [comment]
Dear @lawrence-cj  please add this feature to your official SANA pipeline

File: src/diffusers/models/autoencoders/autoencoder_dc.py (hunk 2 above)
Comment (chenjy2003) [reply src/diffusers/models/autoencoders/autoencoder_dc.py]: Sure. I have modified the code according to your instruction.

File: src/diffusers/models/autoencoders/autoencoder_dc.py (hunk 1 above)
Comment (chenjy2003) [reply src/diffusers/models/autoencoders/autoencoder_dc.py:482]: OK.

**Comment by chenjy2003** [comment]
@elismasilva @a-r-r-o-w I've added the four methods `enable_vae_slicing`, `disable_vae_slicing`, `enable_vae_tiling`, and `disable_vae_tiling` to `SanaPipeline`.

**Comment by elismasilva** [comment]
> @elismasilva @a-r-r-o-w I've added the four methods `enable_vae_slicing`, `disable_vae_slicing`, `enable_vae_tiling`, and `disable_vae_tiling` to `SanaPipeline`.

ok i think you forgot to do the same for this pipeline 
https://github.com/huggingface/diffusers/blob/main/src/diffusers/pipelines/pag/pipeline_pag_sana.py

**Comment by chenjy2003** [comment]
> > @elismasilva @a-r-r-o-w I've added the four methods `enable_vae_slicing`, `disable_vae_slicing`, `enable_vae_tiling`, and `disable_vae_tiling` to `SanaPipeline`.
> 
> ok i think you forgot to do the same for this pipeline https://github.com/huggingface/diffusers/blob/main/src/diffusers/pipelines/pag/pipeline_pag_sana.py

Thanks for pointing out! @lawrence-cj  Could you please double-check whether no other SANA pipelines need to be modified?

**Comment by lawrence-cj** [comment]
We have two pipelines in diffusers. I'm ok with it! Thanks @chenjy2003 for hard work!

**Comment by nitinmukesh** [comment]
Subscribing for it to be merged

**Comment by FurkanGozukara** [comment]
what is difference between pipeline_pag_sana.py and pipeline_sana.py

**Comment by chenjy2003** [comment]
> what is difference between pipeline_pag_sana.py and pipeline_sana.py

@lawrence-cj Here is a question for you.

**Comment by nitinmukesh** [comment]
@FurkanGozukara 

It was answered here
https://github.com/NVlabs/Sana/issues/107#issuecomment-2558488343

**Comment by FurkanGozukara** [comment]
> @FurkanGozukara @chenjy2003
> 
> It was answered here [NVlabs/Sana#107 (comment)](https://github.com/NVlabs/Sana/issues/107#issuecomment-2558488343)

which one is recommended it doesnt tell much

**Comment by nitinmukesh** [comment]
> > @FurkanGozukara @chenjy2003
> 
> which one is recommended it doesnt tell much

Based on my test results PAG quality is 60-70% lower. The question remains why create PAG pipeline or in which specific scenario it will be useful.

**Comment by FurkanGozukara** [comment]
> > > @FurkanGozukara @chenjy2003
> > 
> > 
> > which one is recommended it doesnt tell much
> 
> Based on my test results PAG quality is 60-70% lower. The question remains why create PAG pipeline or in which specific scenario it will be useful.

thanks a lot

**Comment by lawrence-cj** [comment]
> Based on my test results PAG quality is 60-70% lower. The question remains why create PAG pipeline or in which specific scenario it will be useful.

PAG is a function. Anyone can test and try to improve it. We just did the basic work for the community.

**Comment by nitinmukesh** [comment]
> 
> PAG is a function. Anyone can test and try to improve it. We just did the basic work for the community.
Makes sense. So it's primarily for developers to improve or enhance it.

**Comment by bghira** [comment]
this guy Furkan should just be blocked to be honest, creates too much noise and demands a lot from others without thanks

**Comment by lawrence-cj** [comment]
> Makes sense. So it's primarily for developers to improve or enhance it.

Correct.

File: src/diffusers/models/autoencoders/autoencoder_dc.py (hunk 1 above)
Comment (a-r-r-o-w) [reply src/diffusers/models/autoencoders/autoencoder_dc.py]: I've reverted this change because it might be breaking in some cases on second thought. Since we can configure this while using `enable_tiling()`, users should be able to configure it

**Review by a-r-r-o-w — APPROVED** [review]
Thanks! Will merge once tests pass

**Comment by a-r-r-o-w** [comment]
Verified that it works as expected:

```python
import torch
from diffusers import SanaPipeline

pipe = SanaPipeline.from_pretrained(
    "Efficient-Large-Model/Sana_1600M_2Kpx_BF16_diffusers",
    variant="bf16",
    torch_dtype=torch.bfloat16,
)
pipe.to("cuda")
pipe.enable_vae_tiling()

prompt = 'Self-portrait oil painting, a beautiful cyborg with golden hair, 8k'
image = pipe(
    prompt=prompt,
    height=2048,
    width=2048,
    guidance_scale=5.0,
    num_inference_steps=20,
    generator=torch.Generator().manual_seed(42),
)[0]
image[0].save("output.png")
```

image

**Comment by geronimi73** [comment]
Thank you all! Works beautifully 
Something worth mentioning in the docs or model card: the AE's default `tile_sample_min_height` of 512 creates artifacts on large homogenous areas like a blue sky when generating 4k images. Increasing it to 1024 fixed it for the images I looked at and it still runs without OOM on a 3090/4090.

**Comment by FurkanGozukara** [comment]
> `tile_sample_min_height`

can you share your code how do you generate?

**Comment by geronimi73** [comment]
> can you share your code how do you generate?

[notebook](https://github.com/geronimi73/3090_shorts/blob/main/NVIDIA-Sana-4k.ipynb) @FurkanGozukara, if you ever get famous with your genious YouTube videos, make sure to mention me in your nobel prize speech

**Comment by FurkanGozukara** [comment]
> > can you share your code how do you generate?
> 
> [notebook](https://github.com/geronimi73/3090_shorts/blob/main/NVIDIA-Sana-4k.ipynb) @FurkanGozukara, if you ever get famous with your genious YouTube videos, make sure to mention me in your nobel prize speech

thanks a lot. sure if i get famous haha :D

**Comment by bghira** [comment]
can you take the unrelated discussion elsewhere?

**Comment by lawrence-cj** [comment]
> Thank you all! Works beautifully Something worth mentioning in the docs or model card: the AE's default `tile_sample_min_height` of 512 creates artifacts on large homogenous areas like a blue sky when generating 4k images. Increasing it to 1024 fixed it for the images I looked at and it still runs without OOM on a 3090/4090.

How much GPU memory is needed for 3090/4090? I tested on A100, requiring 22GB.

**Comment by geronimi73** [comment]
> How much GPU memory is needed for 3090/4090? I tested on A100, requiring 22GB. 

What I meant what was that generating 4096x4096px images with `tile_sample_min_height=1024` runs on a GPU with 24GB VRAM like a NVIDIA GefForce RTX 3090 or 4090. Sorry for the confusion.

**Comment by lawrence-cj** [comment]
Oh nevermind. Thanks man @geronimi73

**Comment by FurkanGozukara** [comment]
I updated my Gradio APP

4k model now works as low as 8GB GPU VRAM

Diffusers pipeline has amazing improvements 

When all enabled 1k 1024x1024 = 4 GB VRAM GPUs (1)

**Comment by elismasilva** [comment]
> > How much GPU memory is needed for 3090/4090? I tested on A100, requiring 22GB. 
> 
> What I meant what was that generating 4096x4096px images with `tile_sample_min_height=1024` runs on a GPU with 24GB VRAM like a NVIDIA GefForce RTX 3090 or 4090. Sorry for the confusion.

My 3060 Ti 8GB runs great, if you enable cpu offload and vae slicing and vae tiling.
